### PR TITLE
Working implementation of file upload among other things. Fixes #8

### LINKED ISSUE #8: Trigger a file upload.

### PR DESCRIPTION
**Pull request by bschroeter**
In principle working implementation that communicates with ME.org at the Python level. Next will be to implement a CLI for direct access via benchcab.

### CODE REVIEW

**Comment by abhaasgoyal** [comment]
Some small suggestions to additional information in documentation (will look into running aspects as well): 
- Available API calls for the end-user. I'm assuming that they are all in `client.py`, but some of them are a bit unclear, like `login` (which is used by `__init__` but has no indication of being private
- Explanation of workflow - Package dependencies (`requests`), how to build, run and test

File: meorg_client/client.py [hunk 1]
```diff
@@ -25,84 +26,128 @@ def __init__(self, base_url=mcc.MEORG_URL, email=None, password=None):
...
             User password, by default None
         """
-        self.base_url = base_url
-        self.headers = dict()
+
+        # Initialise the mimetypes
+        mt.init()
+
+        # Ensure there is a trailing slash
+        self.base_url = f"{base_url}/" if base_url[-1] != "/" else base_url
```
Comment (abhaasgoyal) [inline meorg_client/client.py:34]: The passed on `base_url` could be empty. An initial check, or using `endswith()` could be some ways.

File: meorg_client/client.py [hunk 2]
```diff
@@ -25,84 +26,128 @@ def __init__(self, base_url=mcc.MEORG_URL, email=None, password=None):
...
+        self.headers = {"Cache-Control": "no-cache", "Pragma": "no-cache"}
         self.last_response = None
 
-        # Automatically login.
+        # Automatically login if credentials are set.
         if email is not None and password is not None:
             self.login(email, password)
```
Comment (abhaasgoyal) [inline meorg_client/client.py:40]: Maybe a warning/log that email/password isn't set?

File: meorg_client/client.py (hunk 1 above)
Comment (abhaasgoyal) [reply meorg_client/client.py:34]: Minor typo in the docstring - Supplying instead of Suppying

File: meorg_client/client.py [hunk 3]
```diff
@@ -25,84 +26,128 @@ def __init__(self, base_url=mcc.MEORG_URL, email=None, password=None):
...
+            Endpoint to be appended to the base URL.
+        **kwargs :
+            Key/value pairs to interpolate into the URL template.
+
+        Returns
+        -------
+        str
+            URL.
+        """
+        return urljoin(self.base_url + "/", endpoint).format(**kwargs)
```
Comment (abhaasgoyal) [inline meorg_client/client.py:134]: Considering that `base_url` should have `/` from 34, not sure adding the / would help

**Review by abhaasgoyal — CHANGES_REQUESTED** [review]

File: meorg_client/client.py [hunk 4]
```diff
@@ -127,51 +172,54 @@ def login(self, email: str, password: str):
...
+            method=mcc.HTTP_POST, endpoint=endpoints.LOGOUT, return_json=False
+        )
 
         # Clear the headers.
-        self.headers = dict()
-
-        return response
+        if response.status_code == 200:
+            self.headers.pop("X-User-Id", None)
+            self.headers.pop("X-Auth-Token", None)
```
Comment (abhaasgoyal) [inline meorg_client/client.py:206]: Maybe have an error case if request wasn't successful (same for `login`). Also checking for 200 response code exactly here (I think creating a smaller function like `success()` would be good?

**Review by abhaasgoyal — CHANGES_REQUESTED** [review]

File: meorg_client/client.py (hunk 1 above)
Comment (bschroeter) [reply meorg_client/client.py:34]: I think a better solution would be to explicitly set the base URL inside the class and set a dev argument. I don't see a situation where the URL is likely to change and if it does we will need to make internal changes anyway.

Will change.

File: meorg_client/client.py (hunk 2 above)
Comment (bschroeter) [reply meorg_client/client.py:40]: Not super necessary. There may be situations where the class will need to be instantiated, modified and then the user would log in. Legitimate use case that doesn't constitute a warning.

File: meorg_client/client.py (hunk 3 above)
Comment (bschroeter) [reply meorg_client/client.py:134]: See comment above.

File: meorg_client/client.py (hunk 4 above)
Comment (bschroeter) [reply meorg_client/client.py:206]: Thanks, I think the class will trigger something anyway with the internal request method - but I will check.

File: meorg_client/client.py (hunk 3 above)
Comment (bschroeter) [reply meorg_client/client.py:134]: This is also just a hammer solution that will force the slash if for some reason it has been omitted somewhere along the way. `urljoin()` has some smarts in it to ignore if not needed.

File: meorg_client/client.py (hunk 4 above)
Comment (bschroeter) [reply meorg_client/client.py:206]: Yes, an exception is raised in `_make_request()` which bubbles up here and constitutes an error.

login/logout are the only time we look for the explicit 200 code. If we end up with many more we will look at implementing your suggestion but at the moment it is likely unnecessary abstraction.

**Comment by bschroeter** [comment]
Documentation will come in a future PR. New commit incoming.

File: meorg_client/client.py (hunk 3 above)
Comment (abhaasgoyal) [reply meorg_client/client.py:137]: Maybe a comment would be suitable here

**Review by abhaasgoyal — APPROVED** [review]
LGTM

**Pull request merged**